components/content/content-editor/content-editor.service.ts

```diff
@@ -0,0 +1,185 @@
+import { Fragment, Mark, Node, NodeType } from 'prosemirror-model';
+import { EditorState, Transaction } from 'prosemirror-state';
+import { EditorView } from 'prosemirror-view';
+import { imageMimeTypes, isImage } from '../../../utils/image';
+import { uuidv4 } from '../../../utils/uuid';
+
+export class ContentEditorService {
+	/**
+	 * Ensures that the last node in the editor doc is a specific node.
+	 */
+	public static ensureEndNode(tr: Transaction, nodeType: NodeType) {
+		if (tr.doc.lastChild && tr.doc.lastChild.type.name !== nodeType.name) {
+			const newNode = nodeType.create();
+			return tr.insert(tr.doc.nodeSize - 2, newNode);
```

Comment:
prosemirror treats the root node as being part of the document itself, so when iterating the document nodes list, the 0 and nth index would be the start and end of the root node. So in order to insert something at the start of the document, we'd insert it at index 1, and to insert at the end we'd do n - 1. The additional -1 (bringing us to a total of -2) is because nodeSize is length based and we need to work with 0 based indexes.